Notes: 
* Answer each question in a separate Jupynet Notebook Cell
* Pleas keep the code in your cells short. 
  * In notebook programming cells are typicaly short to facilitate reading. 
  * If well toughout, most answers in this assignment won're require more than 3 or 4 lines of code. 
* Do no change the list of import, i.e., do not add additional libraries. Those included are the only ones you are allowed to use.
* Add your first and Last name below:

Rexter Delos Santos
    


In [1]:
import pandas as pd
from itertools import product
from collections import Counter
from tqdm.notebook import tqdm
import random
import json

In this assignment you will be working with Corona Virus (SARS-CoV2) data that was obtained from the [National Center for Biotechnology Information](https://www.ncbi.nlm.nih.gov/). You will need two files. The first (`data/coronavirus_info.csv`) is small and is provided in the GitHub Repo. The second  (`data_report.jsonl`) is larger so you will need to download a compressed version, which you will need to uncompress prior to using. You can downlod the second file here:

https://www.dropbox.com/s/qdn67rshygz06ff/data_report.jsonl.gz?dl=0

We start by loading `data/coronavirus_info.csv` (Code provided below)

In [2]:
# CODE PROVIDED -- DO NOT REMOVE
table = pd.read_csv("data/coronavirus_info.csv", low_memory=False)
table = table.drop(["US State", "Host Name", "Host Taxonomy ID", "Sequence Type", "Species Taxonomy Id", "Nuc Completeness", "BioProject", "BioSample"], axis=1)

missing = table["Geo Location"].isnull()
table.loc[missing, "Geo Location"] = ""


table.head(10)

Nucleotide Accession                                     Species Name  \
0          NC_045512.2  Severe acute respiratory syndrome coronavirus 2   
1           OK058807.1  Severe acute respiratory syndrome coronavirus 2   
2           OK058777.1  Severe acute respiratory syndrome coronavirus 2   
3           OK058695.1  Severe acute respiratory syndrome coronavirus 2   
4           OK058662.1  Severe acute respiratory syndrome coronavirus 2   
5           OK058592.1  Severe acute respiratory syndrome coronavirus 2   
6           OK056996.1  Severe acute respiratory syndrome coronavirus 2   
7           OK056909.1  Severe acute respiratory syndrome coronavirus 2   
8           OK056850.1  Severe acute respiratory syndrome coronavirus 2   
9           OK056784.1  Severe acute respiratory syndrome coronavirus 2   

       Virus Genus   Virus Family  \
0  Betacoronavirus  Coronaviridae   
1  Betacoronavirus  Coronaviridae   
2  Betacoronavirus  Coronaviridae   
3  Betacoronavirus  Coronaviridae   
4  Betacoronavirus  Coronaviridae   
5  Betacoronavirus  Coronaviridae   
6  Betacoronavirus  Coronaviridae   
7  Betacoronavirus  Coronaviridae   
8  Betacoronavirus  Coronaviridae   
9  Betacoronavirus  Coronaviridae   

                                   Isolate Name  Nucleotide Length  \
0                                    Wuhan-Hu-1              29903   
1     SARS-CoV-2/human/USA/MA-MASPHL-04825/2021              29801   
2     SARS-CoV-2/human/USA/MA-MASPHL-04790/2021              29771   
3     SARS-CoV-2/human/USA/MA-MASPHL-04700/2021              29820   
4     SARS-CoV-2/human/USA/MA-MASPHL-04651/2021              29798   
5     SARS-CoV-2/human/USA/MA-MASPHL-04499/2021              29802   
6  SARS-CoV-2/human/USA/FL-CDC-QDX27934346/2021              29775   
7  SARS-CoV-2/human/USA/CA-CDC-QDX27909662/2021              29775   
8  SARS-CoV-2/human/USA/FL-CDC-QDX27934406/2021              29763   
9  SARS-CoV-2/human/USA/NY-CDC-QDX28007789/2021              29775   

                     Geo Location Collection Date  
0                     Asia; China         2019-12  
1              North America; USA      2021-07-29  
2              North America; USA      2021-08-10  
3              North America; USA      2021-08-15  
4              North America; USA      2021-08-09  
5              North America; USA      2021-07-26  
6     North America; USA: Florida      2021-08-18  
7  North America; USA: California      2021-08-16  
8     North America; USA: Florida      2021-08-18  
9    North America; USA: New York      2021-08-21

### Q.1

* The location of each of the sequences is recorded under the `Geo Location` column.  How many entries are from Asia?
  * Note that for some records, the `Geo Location` column is missing
  * Display the results using the following format: 
    Asia: XXXX,
    North America': XXXX,
    Europe: XXXX,
    Oceania: XXXX,
    Africa: XXXX,
    South America: XXXX 


In [3]:
# Write your code here
def group(table, index, column):
    location = index
    
    if "Asia" in location:
        return "Asia"
    if "North America" in location:
        return "North America"
    if "Europe" in location:
        return "Europe"
    if "Oceania" in location:
        return "Oceania"
    if "Africa" in location:
        return "Africa"
    if "South America" in location:
        return "South America"

geo_count = table["Geo Location"].value_counts().to_frame("Count")
result = geo_count.groupby(lambda x: group(geo_count, x, "Count"))["Count"].sum()

print(f"Asia: {result['Asia']}, North America: {result['North America']}, Europe: {result['Europe']}, Oceania: {result['Oceania']}, Africa: {result['Africa']}, South America: {result['South America']}")

Asia: 3903, North America: 224014, Europe: 189100, Oceania: 10301, Africa: 1405, South America: 534


* How many entries are from Asia?
    * There are 3903 entries from Asia.

### Q.2
Use the `coronavirus_info.csv` table to count the entries that are from Hawaii. 

In [4]:
# Write your code here
geo_count.loc["North America; USA: Hawaii"]

Count    119
Name: North America; USA: Hawaii, dtype: int64

There are 119 entries from Hawaii

### Q.3

The file `data_report.jsonl` contains the variants of the virus in the DB. This `json` file is a list of records (one per line) for each one of the genomes in the database. Before we work with the large file, we will experiment with a file containing a single record.

The file `single_record.json` contains a single sample record. Use the `JSON library to load the file `single_record.json` into a variable called `sample_vir_record`

1. how many first-level keys does this record have?
  * Do not count nested keys. Only those at the top level should be counted



In [5]:
# Write your code here
sample_vir_record = json.load(open("data/single_record.json"))

len(sample_vir_record)

18

There are 18 first-level keys

### Q.4

Each Covid various in this database is classified according to a system referred to as the Pangolin (Phylogenetic Assignment of Named Global Outbreak LINeages) classification. It is not essential to complete the assignment that you understand this system, but if you're interested in learning more, see:

https://cov-lineages.org/resources/pangolin.html

The Pangolin classification of this sample record is nested within the `virus` key:
```json
{ ...
  "virus": {
              ...
              "pangolinClassification": 
              ...
            }
  ... 
}
```
Write code to extract the classification of this record. The result should be `B.1.1.214`

In [6]:
# Write your code here
sample_vir_record["virus"]["pangolinClassification"]

'B.1.1.214'

We hear in the news about `Alpha`, `Beta`, `Delta` variants of concern and recently the Mu variant as being of interest. Basedon your answer to `Q.3`, you may have been tempted to infer that this virus is of type `Beta` since the first letter is `B`. In fact, this variant of type `Alpha` and is a variant of concern. Although not relevant to this exercise, the rules for naming new variants and the list of known `SARS-CoV-2` are provided here:

https://www.pango.network/how-does-the-system-work/what-are-pango-lineages/

https://cov-lineages.org/lineage_list.html


The following short video is very helpful for understanding what a variant is, how it arises, how it's named, and why some variants are more concerning than others.

https://www.youtube.com/watch?v=B8UEZ9cfgz4

### Q.5

Write Python code that counts all the different kinds of variants in `data_report.jsonl`. 
Because the file is 7GB, you won't likely be able to load it into your laptop's RAM using Python. I encountered this error when trying to open it on my laptop

![](https://www.dropbox.com/s/lieo685pafkgm5e/ram_error.png?dl=1)


It would be easy to extract the data from the pangolinClassification field of each `json` record by reading each line (a record) at a time.

The list of current variants of concern we are interested in are:
      * Alpha (B.1.1.7)
      * Beta (B.1.351, B.1.351.2, B.1.351.3)
      * Delta (B.1.617.2, AY.1, AY.2, AY.3)
      * Gamma (P.1, P.1.1, P.1.2) 

You should get something similar to what follows:
```
Alpha: X
Beta: X
Delta: X
Gamma: X
```
Where `X` represents the counts for relevant variants

In [7]:
# Write your code here
def checkvariant(record):
    try:
        variant = record["virus"]["pangolinClassification"]
        return variant
    except KeyError:
        return None

In [8]:
alpha = ["B.1.1.7"]
beta = ["B.1.351", "B.1.351.2", "B.1.351.3"]
delta = ["B.1.617.2", "AY.1", "AY.2", "AY.3"]
gamma = ["P.1", "P.1.1", "P.1.2"]

variant_count = {
    "alpha": 0,
    "beta": 0,
    "delta": 0,
    "gamma": 0,
}

with open("./data/data_report.jsonl", "r") as vir_records:
    for record in vir_records:
        json_record = json.loads(record)
        variant = checkvariant(json_record)
        
        if variant is None:
            continue
        if variant in alpha:
            variant_count["alpha"] += 1
            continue
        if variant in beta:
            variant_count["beta"] += 1
            continue
        if variant in delta:
            variant_count["delta"] += 1
            continue
        if variant in gamma:
            variant_count["gamma"] += 1
            continue

print(f"Alpha: {variant_count['alpha']}")
print(f"Beta: {variant_count['beta']}")
print(f"Delta: {variant_count['delta']}")
print(f"Gamma: {variant_count['gamma']}")

Alpha: 178423
Beta: 584
Delta: 4392
Gamma: 6632


# Find similar viruses.

It's often useful to compare viruses to study how similar strains are. While sophisticated algorithms to compare a pair of viruses exist, these are typically computationally intensive and cannot be used to carry out a large number of comparisons. 

An alternative, albeit less sensitive, approach consists of comparing word counts (called k-mers, where k is the word size) across genomes.  Suppose we have two viruses X and Y, with the following Genomes.
```
X = "ACGTAGTGCATGTGTAGCTGTGTAGCTGTAC"
Y = "ACTAGTGCATGTGTAGCTCTGTAGCTGATAC"
```

To compare `X` and `Y`, we first vectorize these genomes by marking the presence of words (k-mers) as a boolean value, 0 if absent and 1 if the word is present. This method assumes that similar genomes will have the same words, which makes sense.

This idea, which is referred to as the bag of words model is computationally efficient, making it ideal to vectorize text in big data analytics. Another variant of this model requires replacing the presence and absence by counts for each word.

The code below vectorizes an input DNA sequence intro k-mers of size k=2

In [9]:
# CODE PROVIDED -- DO NOT REMOVE
def get_kmer_2(X):
    DNA = ["A", "C", "G", "T"]
    words_size_2 = ["".join(dna_prod) for dna_prod in product(DNA, DNA)]
    counts = pd.Series([0 for _ in words_size_2], index = words_size_2)
    words_in_X = set([X[i:i+2] for i in range(0, len(X)-1)])
    counts[list(words_in_X)] = 1
    return counts    

In [10]:
# CODE PROVIDED -- DO NOT REMOVE
# X has 3 words of size 2 (AC, CG, GT)
X = "ACGT"
get_kmer_2("ACGT")

AA    0
AC    1
AG    0
AT    0
CA    0
CC    0
CG    1
CT    0
GA    0
GC    0
GG    0
GT    1
TA    0
TC    0
TG    0
TT    0
dtype: int64

The function below takes a dictionary of sequences' counts as a `pandas Series` and prints it using HTML Table, which you might agree is nicer to visualize.

In [11]:
# CODE PROVIDED -- DO NOT REMOVE
def pretty_print_counts(counts_dict):
    list_of_count = [data.to_list() for data in counts_dict.values()]
    list_of_indices = [x for x in counts_dict.keys()]
    list_of_columns = list(counts_dict.values())[0].index.to_list()
    df_single_level_cols = pd.DataFrame(list_of_count,
                                        index=[x for x in counts_dict.keys()],
                                       columns = list_of_columns)    
    return df_single_level_cols 



In [12]:
# CODE PROVIDED -- DO NOT REMOVE

X = "ACGTACGTACGTACGT"
Y = "ACGTACAAACGTACGT"
Z = "TTTTACAAACGTTTTT"

counts_dict = {"X": get_kmer_2(X), "Y": get_kmer_2(Y), "Z": get_kmer_2(Z)}
pretty_print_counts(counts_dict)


AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT
X   0   1   0   0   0   0   1   0   0   0   0   1   1   0   0   0
Y   1   1   0   0   1   0   1   0   0   0   0   1   1   0   0   0
Z   1   1   0   0   1   0   1   0   0   0   0   1   1   0   0   1

### Q.6

Write a function that computes the Jaccard similarity between two feature vectors, A and B. If you recall, Jaccard similarity is computed as:

$$J(A,B) = \frac{A \cap B}{A \cup B}$$

In other words, the number of items shared by `A` and `B` over the set of all items in `A` or `B`.

For example, for `A= get_kmer_2(X)` and Y = get_kmer_2(B) above,

$$
J(A,B) = \frac{4}{6}
$$

Your function should have the following signature: 

`jaccard(A, B)`

Where `A` and `B` are `pandas Series`


Test your function using the code below to make sure it's correct.

In [13]:
# Write you code here
def jaccard(A, B):
    shared = 0
    all_items = 0
    
    for index, value in A.items():
        if value == 0 and B[index] == 0:
            continue
        if value == 1 and B[index] == 1:
            shared += 1
            all_items += 1
            continue
        if value == 0 and B[index] == 1:
            all_items += 1
            continue
        if value == 1 and B[index] == 0:
            all_items += 1
            continue
    return shared/all_items

In [14]:
# TEST PROVIDED -- DO NOT REMOVE
A = get_kmer_2(X)
B = get_kmer_2(Y)
assert jaccard(A, B) == 4/6

### Q.7 

* Compute the jaccard similarity for the pairs of sequences `(X, Y)`, `(X, Z)`, `(Y, Z)`. 


In [15]:
# Write you code here
C = get_kmer_2(Z)

print(f"(X, Y): {jaccard(A, B)}")
print(f"(X, Z): {jaccard(A, C)}")
print(f"(Y, Z): {jaccard(B, C)}")

(X, Y): 0.6666666666666666
(X, Z): 0.5714285714285714
(Y, Z): 0.8571428571428571


### Q.8

The vectors representing the presence and absence of words in both `Y` and `Z` are very similar (Jaccard = 0.85), despite major differences at the DNA level between these two sequences. This is because the words are small -- it is as if you were comparing a history book with a book on Python using words of size 2. It's very likely that both books will contain the same words of size 2. Increasing the size of `k` will produce substantial differences. 

Change the function `get_kmer_2` so that given a sequence `X` and a k-mer size `k`, the function returns a boolean vector of all the words of size `k` in `X`. Cal the function `get_kmer`



The following code can be used to generate all DNA words of size `k`
```pyton
words_size_k = ["".join(prod) for prod in product(*([DNA]*k))]
```

Once done, use the code below to test your function

In [16]:
# Write you code here
# edited get_kmer_2 method
def get_kmer(X, k):
    DNA = ["A", "C", "G", "T"]
    words_size_k = ["".join(prod) for prod in product(*([DNA]*k))]
    counts = pd.Series([0 for _ in words_size_k], index = words_size_k)
    words_in_X = set([X[i:i+k] for i in range(0, len(X)-k+1)])    
    counts[list(words_in_X)] = 1
    return counts    

In [17]:
# TEST PROVIDED -- DO NOT REMOVE
X = "ACGTGATGATTG"

counts = get_kmer(X, k=1)
assert counts.tolist() == [1,1,1,1]


counts = get_kmer(X, k=3)
assert (counts[["ACG", "CGT", "GTG", "TGA", "GAT", "ATG", "ATT", "TTG"]] == 1).sum()  == 8
assert (counts.drop(["ACG", "CGT", "GTG", "TGA", "GAT", "ATG", "ATT", "TTG"]) == 0).sum()  == 56


### Q.9

* Compute the Jaccard similarity for the pairs `(X, Y)`, `(X, Z)`, `(Y, Z)` using `k= 5`


In [18]:
# Write you code here
X = "ACGTACGTACGTACGT"
Y = "ACGTACAAACGTACGT"
Z = "TTTTACAAACGTTTTT"

A = get_kmer(X, k=5)
B = get_kmer(Y, k=5)
C = get_kmer(Z, k=5)

print(f"(X, Y): {jaccard(A, B)}")
print(f"(X, Z): {jaccard(A, C)}")
print(f"(Y, Z): {jaccard(B, C)}")

(X, Y): 0.4
(X, Z): 0.0
(Y, Z): 0.29411764705882354


The appropriate word size varies with the length of the text, with larger words depicting similarity more accurately. However, large values of `k` are:
1. More computationally intensive to compute. With k = 12, there are $4^12 = 16,777,216$ words to compute for each sequence.

2. More likely to skew the distance between fairly similar sequences. For example `k=8`, the Jaccard index between `X` and `Y` is `0`, even though `X` and `Y` have only two mismatching characters. While this is an extreme case due to the fact that X and Y are short, the logic applies to longer sequences and larger values of `k`


![](https://www.dropbox.com/s/rhw5szbiohsqu7w/mismatches.png?dl=1)


In [19]:
# Write you code here
A = get_kmer(X, k=8)
B = get_kmer(Y, k=8)

print(f"(X, Y): {jaccard(A, B)}")

(X, Y): 0.08333333333333333


The code I used is provided as a reference below. The code took 7 hours to complete on a single machine and approximately 12 minutes on a larger server with 72 cores and 1TB of RAM. To parallelize the execution, I split the file into files that contain 1000 sequences each and used GNU Parallel to run each file on a single CPU core.

In [20]:
# CODE PROVIDED FOR ILLUTRATION -- DO NOT REMOVE
# RUNNING LOCALLY MAY TAKE A LONG TIME TO COMPLETE

# k = 8 
# DNA = ["A", "C", "G", "T"]
# words_size_k = ["".join(prod) for prod in product(*([DNA]*k))]

    
# def get_kmer_mod(X):
#     counts = pd.Series([0 for _ in words_size_k], index = words_size_k)
#     words_in_X = set([X[i:i+k] for i in range(0, len(X)-k+1)])
#     counts[list(words_in_X)] = 1
#     return counts   

# def replace_bad_nucs(seq):
#     for character in ['W', 'K', "Y", "M", 'H']:
#         seq = seq.replace(character, 'A') 
        
#     for character in ['R', 'S', 'D', "V", "B"]:
#         seq = seq.replace(character, 'C') 
        
#     seq = seq.replace("N", '') 
    
#     return seq

# all_counts = []
# all_names = []
# with tqdm(total=1000) as pbar:
#     for record in SeqIO.parse("myseq0.fa", 'fasta'):
#         all_names.append(record.id)
#         seq = replace_bad_nucs(str(record.seq))

#         counts = get_kmer_mod(seq)
#         all_counts.append(counts)
#         pbar.update(1)
    
# kmer_counts = pd.DataFrame(all_counts, index = all_names)
# kmer_counts.head()

### Hashing Sequences

We are interested in finding pairs of sequences that are very similar. However, comparing the sequences pairwise is not tractable since it would require carrying out $429282 * (429282 - 1) / 2 = 92_141_303_121$ comparisons.

Instead, we will use the hashing-based approach covered in class. Rather than hashing a sequence over all k-mers, we will only compute the hash for a subset of k-mers. we will repeat the operation n times to avoid that similar sequences are assigned to different bins due to a single, rare mismatch.

This, as discussed in class, is computationally more efficient compared to computing all pairwise sequences. 

### Q.10 

Write a function that takes a `pandas  Series` and a subset of columns and returns the hash computed on the subset of columns. Call this function`hash_on_subset`.

As an example, consider all words with a size of 2 as follows 

|	|AA	|AC	|AG	|AT	|CA | CC| CG| CT| GA| GC| GG| GT| TA| TC| TG| TT|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| A	|0	|1	|0	|0	|0	|0	|1	|0	| 0 |0	|0	| 1 |1  |0	|0	|0  |

```python
hash_on_subset(A, ["AC", "CG", "CT", "GT", "TA"]) 
```

is equivalent to:

```python
hash((1, 1, 0, 1, 1)) == 5085477689562523216
```


In [21]:
# Write you code here
def hash_on_subset(series, subset):
    return hash(tuple(series[subset].values))    

In [22]:
A = get_kmer(X, 2)
hash_on_subset(A, ["AC", "CG", "CT", "GT", "TA"])

5085477689562523216


The method `sample` from the random module `m` words from a list

For example, running:
```python
random.sample( ["A", "C", "G", "T"], 2 )
```
returns
```
['A', 'C']
```
The returned subset may be different for you.

* The code below randomly selects `m=20` k-mers we will use to compare the genomes



In [23]:
random.sample( ["A", "C", "G", "T"], 2 )

['A', 'C']

In [24]:
k=8
DNA = ["A", "C", "G", "T"]
words_size_k = ["".join(prod) for prod in product(*([DNA]*k))]

m=20
subset_kmers = random.sample(words_size_k, m)

subset_kmers

['GTTTTAGA',
 'ACCGTATT',
 'TATGATCG',
 'CCAGCAAT',
 'CGCAATAT',
 'TAACTGGC',
 'ACAGGCCT',
 'TTACACAT',
 'ACGTCATG',
 'GGTGACTT',
 'AGTACCCC',
 'ATGTAAAT',
 'TTTTTGTG',
 'TTTTGGAC',
 'CATAGGGA',
 'GAGCAAGC',
 'TACTGGCT',
 'TTCCGCTA',
 'AAAGCTAT',
 'GTCCTCGA']

In [25]:
all_kmers_df = pd.read_csv("data/all_kmers_10k.csv", low_memory=False)

all_kmers_df.head(3)

Unnamed: 0  AAAAAAAA  AAAAAAAC  AAAAAAAG  AAAAAAAT  AAAAAACA  AAAAAACC  \
0  MT079851.1         1         1         0         0         0         1   
1  OD977007.1         0         0         0         0         0         0   
2  MW796028.1         1         0         0         0         0         0   

   AAAAAACG  AAAAAACT  AAAAAAGA  ...  TTTTTTCG  TTTTTTCT  TTTTTTGA  TTTTTTGC  \
0         0         0         0  ...         0         0         1         1   
1         0         0         0  ...         0         0         1         0   
2         0         0         0  ...         0         0         1         0   

   TTTTTTGG  TTTTTTGT  TTTTTTTA  TTTTTTTC  TTTTTTTG  TTTTTTTT  
0         0         1         0         0         1         1  
1         0         1         0         0         1         1  
2         0         1         0         0         1         1  

[3 rows x 65537 columns]

### Q.12


Apply the function `hash_subset` to all the rows of `all_kmers_df`. The data science (*vectorized*) way to do so is using the `apply` method available on a `pandas DataFrame` instead of using for loops. For example, given a DataFrame `df` such that:

```
df = pd.DataFrame([[1,2,3], [4,5,6]])

```
then 
```
df.apply(max, args=[] axis=1)
```
applies the `max()` function on each row (`axis = 1`). Here, `args` is empty since `max` does not take any additional arguments.

The example below shows how to use `apply` when the function requires additional arguments. In this example, we apply a function that sums all the values of a row and adds to the sum an offset (2 by default)



In [27]:
# EXAMPLE CODE PROVIDED -- DO NOT REMOVE
def add_val_to_sum(x, offset=2):
    return x.sum() + offset
    
df = pd.DataFrame([[1,2,3], [4,5,6]])

print("The sum of rows + an offset of 5 is:")
print(df.apply(add_val_to_sum, args=[5], axis=1))

print("The sum of rows + an offset of 10 is:")
print(df.apply(add_val_to_sum, args=[10], axis=1))
   

The sum of rows + an offset of 5 is:
0    11
1    20
dtype: int64
The sum of rows + an offset of 10 is:
0    16
1    25
dtype: int64


In [28]:
columns = list(all_kmers_df)[1:]

all_kmers_df.apply(hash_on_subset, args=[columns], axis=1)

0      -2011593514066223020
1       -480484593810754780
2        504181857511540197
3      -3155588211435244260
4      -4222097704957565145
               ...         
9983    3659154710393103504
9984    5723540320471716750
9985    5394310158935129352
9986    1428817136028681051
9987    6090634893416330477
Length: 9988, dtype: int64

### Q.13


Use `apply()` to apply `hash_subset` and compute the hash values for all the rows of `all_kmers_df` over `subset_kmers`

* Create a dict by parsing the results to group sequences that yield the same hash under the same bins. Each key in the dict should be a key and each value is a list of sequences that have the same value.

For example, in the dictionary below, X and Y have the same hash value (123456) over a given subset of kmers, whereas Z has a different hash over the same subsets.

```
{"123456": [X,Y], "654321": [Z]}
```


In [35]:
# Write your code here
hash_dict = {}
hash_subset = all_kmers_df.apply(hash_on_subset, args=[subset_kmers], axis=1)

In [75]:
#hash_subset_df = pd.DataFrame(data=[hash_subset.values, all_kmers_df['Unnamed: 0'].values])
#hash_subset_df.apply(create_dict, args=[hash_dict], axis=0)

dna = all_kmers_df['Unnamed: 0']

for index, value in hash_subset.iteritems():
    try:
        hash_dict[value].append(dna[index])
    except KeyError:
        hash_dict[value] = [dna[index]]

hash_dict

{7690099599645525206: ['MT079851.1',
  'OD977007.1',
  'MW796028.1',
  'MW559193.1',
  'OU031839.1',
  'MW030232.1',
  'MZ553971.1',
  'MZ850453.1',
  'MZ371569.1',
  'MZ001071.1',
  'MZ592007.1',
  'MZ002141.1',
  'MW908942.1',
  'MZ002723.1',
  'MZ069458.1',
  'MZ141510.1',
  'OU031807.1',
  'OU030348.1',
  'OD984828.1',
  'OD977249.1',
  'OD968949.1',
  'OD951702.1',
  'OD949720.1',
  'OD947749.1',
  'OD944351.1',
  'OD941005.1',
  'OD937713.1',
  'OD936233.1',
  'OD935257.1',
  'OD932298.1',
  'OD929645.1',
  'OD926714.1',
  'OD925457.1',
  'OD923955.1',
  'OD922016.1',
  'OD920955.1',
  'OD919999.1',
  'OD918200.1',
  'OD916595.1',
  'OD915123.1',
  'OD913898.1',
  'OD913009.1',
  'OD911876.1',
  'OD910799.1',
  'OD909996.1',
  'OD909201.1',
  'OD908628.1',
  'OD907986.1',
  'OD907378.1',
  'OD906189.1',
  'OD905491.1',
  'OD904712.1',
  'OD904011.1',
  'OD903451.1',
  'OD902842.1',
  'OD902201.1',
  'OD901609.1',
  'OD900993.1',
  'OD900419.1',
  'OD899843.1',
  'OD899264.1',
  '

### Q. 15

Here we use the presence and absence of words, i.e., a vector of booleans, to encode a sequence. The problem with this approach is that it considers the sequences to be identical, even if their word counts differ substantially. For example, given the sequence `X`, `Y` and `Z` as follows
```
X = ATAGATAGATAGATAGATT
Y = ATAGATAGATAGATAGATT
Z = ATAGATTTTTTTTTTTTTT
```
With k =2, all three sequences mach on their vector of word presence/absense.

In [30]:
X = "ATAGATAGATAGATAGATT"
Y = "ATAGATAGATAGATAGATT"
Z = "ATAGATTTTTTTTTTTTTT"

counts_dict = {"X": get_kmer(X, k=2), "Y": get_kmer(Y, k=2), "Z": get_kmer(Z, k=2)}
pretty_print_counts(counts_dict)

AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT
X   0   0   1   1   0   0   0   0   1   0   0   0   1   0   0   1
Y   0   0   1   1   0   0   0   0   1   0   0   0   1   0   0   1
Z   0   0   1   1   0   0   0   0   1   0   0   0   1   0   0   1

* Comparing these sequences based on word counts, X is much more similar to Y than it is to Z

In [31]:
def get_kmer_counts(X, k):
    DNA = ["A", "C", "G", "T"]
    words_size_k = ["".join(prod) for prod in product(*([DNA]*k))]
    counts = pd.Series([0 for _ in words_size_k], index = words_size_k)
    counts_words_in_x = Counter([X[i:i+k] for i in range(0, len(X)-k+1)])
    counts.update(counts_words_in_x)
    return counts    


counts_dict = {"X": get_kmer_counts(X, k=2), "Y": get_kmer_counts(Y, k=2), "Z": get_kmer_counts(Z, k=2)}
pretty_print_counts(counts_dict)

AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT
X   0   0   4   5   0   0   0   0   4   0   0   0   4   0   0   1
Y   0   0   4   5   0   0   0   0   4   0   0   0   4   0   0   1
Z   0   0   1   2   0   0   0   0   1   0   0   0   1   0   0  13

* Given an example (vectors) to justify why hashing is not ideal with counts.
  * Use any means you think are useful to illustrate your point (e.g.: figure, simulation (yes, please!))
 
* Describe how the random project approach discussed in class can help solve the issue discussed
  * Use code to illustrate how random projection works in the following example.
    * I.e., provide code to provide an example where `X` and `Y` are assigned to the same bin and `Y` is assigned to a different bin.
    * You can choose any vector values as needed 
 
```python
X = [1,2]
Y = [2,2]
Z = [5,1]
```

In [32]:
### Write your code here